src/pdf_beamtime/src/pdf_beamtime_server.cpp

```diff
@@ -137,6 +156,82 @@ std::vector<moveit_msgs::msg::CollisionObject> PdfBeamtimeServer::create_env()
   return all_obstacles;
 }
 
+void PdfBeamtimeServer::new_obstacle_service_cb(
+  const std::shared_ptr<NewObstacleMsg::Request> request,
+  std::shared_ptr<NewObstacleMsg::Response> response)
+{
+  try {
+    // Adds the new obstacle name to the existing list
+    auto obj_param = node_->get_parameters({"object_names"})[0].as_string_array();
+    obj_param.push_back(request->name);
+    node_->set_parameters(
+      {rclcpp::Parameter(
+          "object_names",
+          obj_param)});
+
+    // Add new parameters one by one
+    node_->declare_parameter("objects." + request->name + ".type", request->type);
+    node_->declare_parameter("objects." + request->name + ".x", request->x);
+    node_->declare_parameter("objects." + request->name + ".y", request->y);
+    node_->declare_parameter("objects." + request->name + ".z", request->z);
+    node_->declare_parameter("objects." + request->name + ".w", request->w);
+    node_->declare_parameter("objects." + request->name + ".h", request->h);
+    node_->declare_parameter("objects." + request->name + ".d", request->d);
+    node_->declare_parameter("objects." + request->name + ".r", request->r);
+
+    // Return response results
+    response->results = "Success";
+  } catch (const std::exception & e) {
+    std::cerr << e.what() << '\n';
+    response->results = "Failure";
+  }
+  // Update the whole environment
+  planning_scene_interface_.applyCollisionObjects(create_env());
+}
+
+void PdfBeamtimeServer::update_obstacles_service_cb(
+  const std::shared_ptr<UpdateObstaclesMsg::Request> request,
+  std::shared_ptr<UpdateObstaclesMsg::Response> response)
+{
+  // Update the existing parameter
+  auto results = node_->set_parameter(
+    rclcpp::Parameter(
+      "objects." + request->name + "." + request->property,
+      request->value));
+
```

Comment:
This is an interesting design choice. The implication is that I don't change an object, but a single attribute of the object.

Reply:
Service was changed to send out multiple attributes and their values in an array, and those attributes are changed to the specified values in a loop.
https://github.com/maffettone/erobs/blob/ed89fd89b78dfd1d51e227f933d04c9d4588b283/src/pdf_beamtime/src/pdf_beamtime_server.cpp#L196-L201